enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,320 @@
+---
+
+# Improve CoreDNS Integration with EgressFirewall
+
+## Summary
+
+This enhancement improves the integration of CoreDNS with EgressFirewall. With this improved
+integration, EgressFirewall will be able to better support DNS names whose IPs change dynamically
+and also will be able to provide support for wildcard DNS names.
+
+## Motivation
+
+Currently, EgressFirewall (OVN-K master) does a DNS lookup of a DNS name based on a default TTL or the
+TTL of the previous lookup. OVN-K master then updates the underlying ``AddressSet`` for the DNS name
+referenced by the corresponding ACL rule of the EgressFirewall. However, if a pod, belonging to the same
+Namespace as that of the EgressFirewall, does a DNS lookup and is able to get the response before the OVN-K
+master then the pod will be incorrectly allowed/denied access to the DNS name. The integration between
+CoreDNS and EgressFirewall needs to be improved to avoid such a scenario.
+
+If an administrator wants to specifically allow/deny access to all subdomains then currently the
+administrator has to add all subdomains in the EgressFirewall rules. This becomes difficult when
+subdomains are dynamically added/removed as each one has to be added/removed individually from the
+EgressFirewall rules. Currently, wildcard DNS names are not supported in EgressFirewall. However,
+even if the support is added to EgressFirewall, the integration between CoreDNS and EgressFirewall
+needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.
+
+
+### User Stories
+
+* As an OpenShift cluster administrator, I want to add regular DNS names to EgressFirewall rules, so that I can allow/deny
+access to them even if the IPs associated with the corresponding DNS records change dynamically.
+* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
+allow/deny access to all the subdomains belonging to the wildcard DNS names.
+* As an OpenShift engineer, I want to add a new Custom Resource, so that IPs and TTLs of the DNS names which are used
+in EgressFirewall rules can be tracked.
+* As an OpenShift engineer, I want to add a new plugin to CoreDNS, so that DNS lookups of DNS names used in EgressFirewall
+rules can be inspected and the current IPs and TTLs can be tracked in the corresponding new CR.
+* As an OpenShift engineer, I want to modify Cluster DNS operator, so that CoreDNS can be deployed with the
+new plugin enabled.
+
+### Goals
+
+* Support update of the ``AddressSet`` referenced by the ACL rule corresponding to an EgressFirewall if IPs associated
+with the corresponding DNS names change dynamically.
+* Support usage of wildcard DNS names in EgressFirewall rules.
+* Create CRs for each unique DNS name used in the EgressFirewall rules and use the CRs to track the current
+IPs and the corresponding TTL information.
+
+### Non-Goals
+
+* Support additional DNS resolution functionality in the new CoreDNS plugin.
```

Comment:
What I meant here is the new plugin will only inspect the response of DNS resolution by other existing plugins. It will not do any DNS lookup (resolution) by itself.